support/Python/DirectoryStructure.py

```diff
@@ -126,3 +126,99 @@ def list_segments(segments_dir: Union[str, Path]) -> List[Segment]:
         return []
     matches = map(Segment.match, segments_dir.iterdir())
     return sorted(match for match in matches if match)
+
+
+@dataclass(frozen=True, order=True)
+class PipelineStep:
+    """A complete simulation iteration
+
+    Represents a full simulation run. Multiple iterations can exist,
+    organized in a directory structure like this:
+
+    ```
+    BASE_DIR/
+        001_InitialData/
+            Segments/
+                Segment_0000/
+                    Checkpoints/
+                    InputFile.yaml
+                    Submit.sh
+                Segment_0001/
+                    ...
+        002_Inspiral/
+            ...
+    ```
+
+    Note: "InitialData" and "Inspiral" are examples, any name can be used.
+    """
+
+    path: Path
+    id: int
+    label: str
+
+    NAME_PATTERN = re.compile(r"(\d+)_(.+)")
+    NUM_DIGITS = 3
+
+    @classmethod
+    def first(cls, directory: Path, label: str) -> "PipelineStep":
```

Comment:
Change `Path` -> `Union[str, Path]` so we can pass in the directory either as a string or as a `Path` object. Then below change `directory / name` -> `Path(directory) / name` to make sure a string is converted to a Path before appending the name.

(optional) also change this in the `Segment` class above if you get the chance